src/odemis/gui/cont/temperature.py

```diff
@@ -0,0 +1,119 @@
+        self._len_temp_q = len_history
+
+        self._time_last_warning = 0  # time of the last warning
+        self._time_last_speed_warning = 0
+
+        val = self.thermostat.temperature.value
+        self.temperature_history = collections.deque(self._len_temp_q * [val], self._len_temp_q)
+        self._update_temperature_timer = RepeatingTimer(self._temperature_sample_period, self._updateTemperature)
+        self._update_temperature_timer.start()
+
+    def _updateTemperature(self):
+        new_temperature = self.thermostat.temperature.value
+        self.temperature_history.append(new_temperature)
+
+        # Get boundaries
+        safe_range = self.thermostat.getMetadata().get(md.MD_FAV_POS_RANGE, None)
+        safe_speed = self.thermostat.getMetadata().get(md.MD_SAFE_SPEED_RANGE, None)
+
+        # Calculate a linear regression of the temperature history to determine the speed
+        t = numpy.linspace(-(self._len_temp_q - 1) * self._temperature_sample_period, 0, self._len_temp_q)
+        try:
+            speed, _ = numpy.polyfit(t, list(self.temperature_history), 1)
+        except:
+            logging.exception("Could not compute the rate of change of temperature.")
+            return
+
+        # logging.debug(u"Temperature: %f °C, speed is %f °C / s", new_temperature, speed)
+
+        # determine setting the warning
+        if safe_range is not None and (new_temperature > safe_range[1] or new_temperature < safe_range[0]):
+            # temperature is out of range
+            # change colour to red
+            wx.CallAfter(self.temperature_label.SetForegroundColour, wx.RED)
+            text = u"Temperature {temp:.2f}°C is outside of the safe operating range!".format(temp=new_temperature)
+            logging.warning(text)
+            # display messagebox warning, but only once per minute
+            if time.time() - self._time_last_warning > 60:
+                wx.CallAfter(wx.MessageBox, text, "Warning", wx.ICON_WARNING)
+                self._time_last_warning = time.time()
+
+        elif safe_speed is not None and (speed > safe_speed[1] or speed < safe_speed[0]):
+            # temperature is changing at an unsafe speed
+            # change colour to yellow
+            wx.CallAfter(self.temperature_label.SetForegroundColour, wx.YELLOW)
+            text = u"Temperature {temp:.2f}°C is changing at an unsafe rate of {speed:.2f}°C/s!".format(temp=new_temperature, speed=speed)
+            logging.warning(text)
+            # display messagebox warning, but only once per minute
+            if time.time() - self._time_last_speed_warning > 60:
+                wx.CallAfter(wx.MessageBox, text, "Warning", wx.ICON_WARNING)
+                self._time_last_speed_warning = time.time()
+
+        else:
+            # temperature is normal. Set the colour back to normal
+            wx.CallAfter(self.temperature_label.SetForegroundColour, wx.Colour(20, 20, 20))
```

Comment:
Do not hard-code the colour here. You could use a constant, but I think it's easier/better to store the foreground at init, and restore it here.